### DISCUSSION TITLE
Cannot save ComfyUI integration, model list doesn't get populated

### DISCUSSION THREAD

**Discussion by hartmark** [open-webui/open-webui#4571] [General]
# Bug Report

## Installation Method

* latest docker images of ollama/ollama:rocm and ghcr.io/open-webui/open-webui:main
* I have ComfyUI installed so it listens on localhost
* I have in open-webui added a host entry host.docker.internal -> host-gateway so I can access ComfyUI
* I have checked the box "Enable Dev mode Options" in ComfyUI

## Environment
open-webui: v0.3.12

- **Operating System:** arch linux
- **Browser (if applicable):** firefox

**Confirmation:**

- [x] I have read and followed all the instructions provided in the README.md.
- [x] I am on the latest version of both Open WebUI and Ollama.
- [x] I have included the browser console logs.
- [x] I have included the Docker container logs.
- [x] I have provided the exact steps to reproduce the bug in the "Steps to Reproduce" section below.

## Expected Behavior:

I don't seem to be able to connect to ComfyUI

## Actual Behavior:

I get no models populated so I cannot save my settings.

## Description

**Bug Summary:**
Cannot save ComfyUI integration

## Reproduction Details

**Steps to Reproduce:**

1. I go to http://localhost:3122/admin/settings
2. click **images**
3. I select `ComfyUI` under **Image Generation Engine**
4. I enter `http://docker.host.internal` under **ComfyUI Base URL**
5. I click **off** to enable the connection
6. I get this error message:
![image](https://github.com/user-attachments/assets/8f01e677-95db-4dbf-b6ba-73a9602fb67a)

## Logs and Screenshots

**Browser Console Logs:**
```
Message received from iframe: 
Object { source: "react-devtools-content-script", hello: true }
[Chat.svelte:199:12](http://localhost:3122/src/lib/components/chat/Chat.svelte)
Message received from iframe: 
Object { source: "react-devtools-content-script", hello: true }
[Chat.svelte:199:12](http://localhost:3122/src/lib/components/chat/Chat.svelte)
XHRGET
http://localhost:3122/images/api/v1/models
[HTTP/1.1 400 Bad Request 4ms]

Object { detail: "Something went wrong :/\nInvalid URL '/object_info': No scheme supplied. Perhaps you meant [https:///object_info](https://object_info/)?" }
[index.ts:358:11](http://localhost:3122/src/lib/apis/images/index.ts)
Message received from iframe: 
Object { source: "react-devtools-content-script", hello: true }
[Chat.svelte:199:12](http://localhost:3122/src/lib/components/chat/Chat.svelte)
Message received from iframe: 
Object { source:
```

**Docker Container Logs:**
nothing is printed here

**Screenshots/Screen Recordings (if applicable):**
![image](https://github.com/user-attachments/assets/1f5f12a9-3b37-4c08-86f0-2f5b09cf28c4)

**Comment by silentoplayz** [comment]
1. In Open WebUI, navigate to the Image Generation Settings page.
2. Set the `Image Generation Engine` option to `ComfyUI`.
3. Enable the `Image Generation (Experimental)` toggle by setting it to `On`.
4. Ensure that ComfyUI is running, as it is required to verify the connection.
5. Verify the connection by checking the `ComfyUI Base URL`.
6. Once the connection is verified, select a default model from the `Select a model` dropdown for the `Set Default Model` option. This option points to the `/models/checkpoints` directory of your running ComfyUI installation.
7. Save your changes on the Image Generation Settings page to apply the new configuration.

**Comment by hartmark** [comment]
It's step 6 I cannot do. I get the error above and cannot continue. What is the full API call it uses to access the list of models in comfyUI?

**Reply by silentoplayz** [reply]
> It's step 6 I cannot do. I get the error above and cannot continue. What is the full API call it uses to access the list of models in comfyUI?

I'm not able to reproduce the error. If the `ComfyUI Base URL` were invalid upon verifying the connection within Open WebUI, you should've saw `Oops! The URL you provided is invalid. Please double-check and try again.` as an error message. We may need more context about the issue. The `ComfyUI Base URL` would be the URL you use to access the ComfyUI interface.

**Reply by cbai970** [reply]
I am working through this workflow , and I can't seem to validate the URL 
http://127.0.0.1:8188/ works in my browser, but when i plug it into webui , it says invalid. 

I suspect this is an issue with webui being inside docker and it being the wrong external URL.

**Reply by silentoplayz** [reply]
> I am working through this workflow , and I can't seem to validate the URL http://127.0.0.1:8188/ works in my browser, but when i plug it into webui , it says invalid.
> 
> I suspect this is an issue with webui being inside docker and it being the wrong external URL.

Try `http://host.docker.internal:8188` if ComfyUI is on your local computer and if Open WebUI is being ran via Docker Desktop.

**Reply by cbai970** [reply]
and now I run into what I think is hartmark's original problem, it wont populate models. Im gonna guess I have to pull that into comfy...

**Comment by hartmark** [comment]
I have ComfyUI on http://localhost:80

open-webui runs in a docker-compose and has this configuration:
```
  ollama-webui:
    image: ghcr.io/open-webui/open-webui:main
    container_name: ollama-webui
    ports:
      - "3122:8080"
    volumes:
      - ./ollama-webui:/app/backend/data
    extra_hosts:
      - "host.docker.internal:host-gateway"
    restart: unless-stopped
```

I have connected to the docker container, installed ping:

```
apt-get update && apt-get install -y iputils-ping
root@a6557b1c3a10:/app/backend# ping host.docker.internal
PING host.docker.internal (172.17.0.1) 56(84) bytes of data.
64 bytes from host.docker.internal (172.17.0.1): icmp_seq=1 ttl=64 time=0.052 ms
64 bytes from host.docker.internal (172.17.0.1): icmp_seq=2 ttl=64 time=0.029 ms
64 bytes from host.docker.internal (172.17.0.1): icmp_seq=3 ttl=64 time=0.041 ms
64 bytes from host.docker.internal (172.17.0.1): icmp_seq=4 ttl=64 time=0.037 ms
```

In image settings I added:
`http://host.docker.internal:80/`
as **ComfyUI Base URL**

* I then click save.
* then I click the `Off` to enable the integration.
* Then I get the error I had above
`Object { detail: "Something went wrong :/\nInvalid URL '/object_info': No scheme supplied. Perhaps you meant [https:///object_info](https://object_info/)?" }`

**Comment by hartmark** [comment]
I'm able to use curl to add images to be generated. I think it would be nice to be able to upload custom workflows as I need to do a workflow like this to get schnell working using GGUF so I get less VRAM-usage.

I used this guide to get a json file for my workflow:
https://blog.xentoo.info/2023/07/22/comfyui-batch-run-from-command-line-with-api/

And here's the endpoint docs:
https://docs.comfy.org/essentials/comms_routes

[flux-api.json](https://github.com/user-attachments/files/16631073/flux-api.json)

Then I can just modify the placeholder REPLACE_ME and add it to ComfyUI queue by this command:
`PROMPT_ID=$(sed 's/REPLACE_ME/A futuristic city skyline at night/' flux-api.json | curl -s -X POST --data @- http://localhost/prompt | sed -n 's/.*"prompt_id": "\([a-f0-9\-]\+\)".*/\1/p')`

Then I can use this command to check the status of the render:
`curl -s -X GET "http://localhost/queue" | grep "$PROMPT_ID"`

The result would be empty if the item is no longer in queue.

Then I extract the filename of the generated image with this command
`FILENAME=$(curl -s -X GET "http://localhost/history/$PROMPT_ID" | jq -r '.["'$PROMPT_ID'"].outputs["13"].images[0].filename')`

Then lastly I fetch the image with this command:
`curl -s "http://localhost/view?filename=$FILENAME" | catimg - -w 200`

would it be possible to do an implementation like this instead? Or to have an alternative solution to make it more flexible.

**Comment by hartmark** [comment]
It was listening on all interfaces, just that ufw stopped it. I thought local docker access was allowed with ufw.

Just running this command on docker host fixed my issue:
`sudo ufw allow 80/tcp comment sd-web`

**Comment by hartmark** [comment]
Is it possible to get support to just have your own workflow sent to ComfyUI?

**Reply by silentoplayz** [reply]
> Is it possible to get support to just have your own workflow sent to ComfyUI?

There is a PR open for that - https://github.com/open-webui/open-webui/pull/4638

**Comment by hartmark** [comment]
Cool, I'll keep watching that PR to be merged then

**Reply by silentoplayz** [reply]
> Cool, I'll keep watching that PR to be merged then

Relevant - https://github.com/open-webui/open-webui/pull/4638#issuecomment-2301967129

**Comment by fgheorghe** [comment]
For me, the issue is coming from:

retrieval.py -> get_web_loader -> utils.py -> validate_url

If the search results contain a URL that 404's or is otherwise invalid, then the whole search will fail.

Perhaps a better approach may be to add an option to filter out invalid URLs as opposed to failing the prompt request all together?
